### ISSUE TITLE
Crash when reading MFT

### ISSUE THREAD

**Issue by Aaron-Ritter-IDEXX** [dr-anoroc/rawccopy#5]
I'm getting a crash on one drive with a valid filesystem; I've run chkdsk on the drive, no errors found. This happens when trying to copy any file, including the MFT.

It's happening in `mft_record AttributesForFile(execution_context context, uint64_t mft_ref, uint32_t attribute_mask)` while handling a `ATTR_ATTRIBUTE_LIST`. It calls `BytesFromAttributeList`, that calls `GetMFTRecord` with an index of 15, which chokes on `BlockFromRunList` with block_index of 15.

At this point, we're getting beyond my ability to follow along. Near as I can tell, the problem winds up being that the pointer `run_list` is 0, which means the `context->mft_index` that was passed in is 0. That then fails to dereference in `utarray_front(run_list)`.

I've uploaded a copy of the MFT in question, copied with another tool like this one: https://drive.google.com/file/d/1qLeObUCrLXmvmYJ6OVxjth-oIGRvb-5u/view?usp=sharing

**Comment by dr-anoroc** [comment]
This looks really bad. Couple of questions/comments:
1. `context->mft_index` can never be NULL. That is to say, if you successfully get out of the original set-up and get back a complete `context`, `mft_index` cannot be NULL. In theory it could potentially contain an empty run list, which would be equally bad, but would not crash the program. So this would point to another root cause, meaning that something corrupts the `context`, setting `mft_index` to NULL. Doing a debug and constantly looking at `context->mft_index` to see which operation corrupts it, is a very likely pointer to the real cause of the problem.
2. Your reference to index 15: am I to understand that whatever copy operation you launch, you end up calling `GetMFTRecord` with index 15? That would be very strange, because 15 is one of these index values where no user data is stored. Normally regular file records start at MFT value 27.
3. The file you've uploaded is just the MFT and not a complete disk image right? If so, would it be at all possible to upload a complete disk image? I can understand that this is a bit sensitive (depending on the content), but it would be a great help of course.

**Comment by Aaron-Ritter-IDEXX** [comment]
1) Indeed, we haven't gotten out of the initial setup yet. I see I left out that detail; this happens during the line `execution_context cont = SetupContext(argc, argv);` of main.c.
Inside there, it calls `if (!SetUpMFTIndex(result))`.
That calls `mft_record mft = AttributesForFile(context, MASTER_FILE_TABLE_NUMBER, AttrTypeFlag(ATTR_DATA));`.
From there it's looping through `for (uint32_t attrib_offs = MFTRecPtr(rec)->attrs_offs; ;)`, and the second loop enters the `if (attr->type == ATTR_ATTRIBUTE_LIST)` condition.
That enters `bytes extra_at = BytesFromAttributeList(context, attr, mft_ref);`.
That enters the loop `for (uint64_t* recID = utarray_front(attrs); recID; recID = utarray_next(attrs, recID))`, and I believe it was the first execution of that loop called `bytes rec = GetMFTRecord(context, (*recID) & 0x0000FFFFFFFFFFFF);` with a value of 15 (I didn't catch what the original *recID value is).
I figured this out by adding printf and fflush lines everywhere. How classic, I haven't done that kind of debugging in a long time.

2) Indeed, it's during the initial context setup, so it's every file. I guess that means it's MFT metadata?

3) The file is just a copy of the MFT, yes. There are apparently parser programs that can turn it into something more human-readable. Unfortunately, I cannot upload the disk image; this is someone else's machine, and while I figured the MFT was sufficiently safe, anything involving file content can't be allowed. However, I can get remotely connected again and can run debug builds or whatever you'd like.

**Comment by dr-anoroc** [comment]
OK, I think I see what's happening.

Could you please try the following to confirm:
In the file mft.c, at the start of the function `GetMFTRecord`, replace the following snippet: 

```
if (index == 0)
        //This is the bootstrap case, where we simply need to read record 0:
        result = ReadImageBytes(context->cr, context->boot->logical_clust_num_for_mft *
			  BytesPerCluster(context->boot), MFTRecordSize(context->boot));
```

by

```
if (index < 27)
        //This is the bootstrap case, where we simply need to read record 0:
	result = ReadImageBytes(context->cr, context->boot->logical_clust_num_for_mft *
		BytesPerCluster(context->boot) + MFTRecordSize(context->boot) * index, MFTRecordSize(context->boot));
```
If this fixes the issue, I'll include it in the next check-in.

**Comment by Aaron-Ritter-IDEXX** [comment]
That seems to have done the trick for this machine. Our QA person is out this week, and I'm out next week, so I won't get to try a general deployment until after that, but I'd call this a success.

**Comment by dr-anoroc** [comment]
I've been studying this a bit further and it all seems to add up. Even if there is no official NTFS spec out there, you can piece it together from various sources. In this case it's nicely explained [here](https://en.wikipedia.org/wiki/NTFS) in the section on the MFT and the Metafiles in it: numbers 0 to 11 have specific names and roles. However, numbers 12 to 23 are:
 
_"Reserved for $MFT extension entries. Extension entries are additional MFT records that contain additional attributes that do not fit in the primary record. This could occur if the file is sufficiently fragmented, has many streams, long filenames, complex security, or other rare situations."_

So they are sort of an overflow buffer in case $MFT 0 is so complicated that it needs more space. That's exactly what's happening in your case (you can even see it in the file you uploaded): $MFT file 0 flows over into file MFT file 15. It's all within specs, just rather exceptional.

This brings me back to your QA reference. I have been constructing a set of test cases to automate testing (this is the reason you haven't seen much progress on the other two issues) and it would be good to include this case in such a test set. Obviously getting a copy of the disk was utopic, so second best would be to get a description of the disk, ie to find what the owner of the disk did to cause this situation. In any case, I think I'm going to play around a bit with disk images to see if I can force them to exhibit the same characteristics. My first priority, however, is to get a patch out for the three open issues.

**Comment by dr-anoroc** [comment]
Fixed in 0.1.7
